File: src/expr.rs
use crate::{location::{Wrapper, Located}, pattern::LocPattern, types::Type0};

#[derive(Debug, Clone)]
pub(crate) enum Expr<P, T> {
    Char(char),
    Int(i64),
    Float(f64),
    Bool(bool),
    Tuple(Vec<T>),

    Binding(String),

    Const(usize, Vec<T>),

    App(T, Vec<T>),
    Abs(Vec<String>, T),

    // let 0 = 1 in 2
    Let(String, T, T),
    Cond(T, T, T),

    // match 0 with 1.0 if 1.1 -> 1.2
    Match(T, Vec<(P, T, T)>),
}

#[derive(Debug, Clone)]
pub(crate) struct LocExprTP0<'a>(
    pub(crate) Located<'a, (Expr<LocPattern<'a>, Box<LocExprTP0<'a>>>, Option<Type0>)>,
);

impl<'a> LocExprTP0<'a> {
    pub(crate) fn map<F>(self, f: F) -> Self
    where
        F: Fn(
            (Expr<LocPattern<'a>, Box<LocExprTP0<'a>>>, Option<Type0>),
        ) -> (Expr<LocPattern<'a>, Box<LocExprTP0<'a>>>, Option<Type0>),
    {
        LocExprTP0(self.0.map(f))
    }
}

#[derive(Debug, Clone)]
pub(crate) struct LocExprTF0<'a>(
    pub(crate) Located<'a, (Expr<LocPattern<'a>, Box<LocExprTF0<'a>>>, Type0)>,
);

impl<'a> LocExprTF0<'a> {
    pub(crate) fn map<F>(self, f: F) -> Self
    where
        F: Fn(
            (Expr<LocPattern<'a>, Box<LocExprTP0<'a>>>, Option<Type0>),
        ) -> (Expr<LocPattern<'a>, Box<LocExprTP0<'a>>>, Option<Type0>),
    {
        LocExprTF0(self.0.map(f))
    }
}

File: app/Expr.hs
module Expr ( Expr(..), Branch(..), LocExprT, LocBranchT ) where

import Data.Text
import Literal ( Literal )
import Location
import Pattern
import Type

data Expr a = Literal { ann :: a, lit :: Literal }
            | Tuple { ann :: a, els :: [Expr a] }
            | Var { ann :: a, v :: Text }
            | Const { ann :: a, k :: Int, cargs :: [Expr a] }
            | App { ann :: a, f :: Expr a, args :: [Expr a] }
            | Abs { ann :: a, fargs :: [(Text, a)], e :: Expr a }
            -- let x = y in z
            | Let { ann :: a, x :: (Text, a), y :: Expr a, z :: Expr a }
            | Cond { ann :: a, ei :: Expr a, et :: Expr a, ee :: Expr a }
            | Match { ann :: a, e :: Expr a, branches :: [Branch a] }

data Branch a = Branch a (Pattern a) (Maybe (Expr a)) (Expr a)

type LocBranchT = Branch LocAnnT

type LocExprT = Expr LocAnnT
